atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -659,25 +689,14 @@ private ResultWithToken getTimestampsByCell(
             ColumnSelection columns,
             long timestamp,
             long batchSize,
-            boolean isReverse) {
+            boolean isReverse,
+            Token token) {
         try (ClosableIterator<AgnosticLightResultRow> iterator = table
-                .getAllRows(rows, columns, timestamp, false, isReverse)) {
-            ResultWithToken result = new ResultWithToken();
-            SetMultimap<Cell, Long> rowBuffer = HashMultimap.create();
-            while (iterator.hasNext() && result.size() + rowBuffer.size() < batchSize) {
-                AgnosticLightResultRow cellResult = iterator.next();
-                result.storePreviousRowIfComplete(cellResult, rowBuffer);
-                Cell cell = Cell.create(cellResult.getBlob("row_name"), cellResult.getBlob("col_name"));
-                long ts = cellResult.getLong("ts");
-                rowBuffer.put(cell, ts);
-            }
-            if (iterator.hasNext()) {
-                AgnosticLightResultRow cellResult = iterator.next();
-                result.storePreviousRowIfComplete(cellResult, rowBuffer);
-            } else {
-                result.storeRow(rowBuffer);
-                result.finishedRange = true;
-            }
+                .getAllRows(rows, columns, timestamp, false, DbReadTable.Order.fromBoolean(isReverse))) {
+            ResultWithToken result = ResultWithToken.create(iterator);
+            result.moveForward(token);
```

Comment:
It looks like ResultWithToken must be used in this way (calling `create`, `moveForward`, `getBatchOfTimestamps`, and `checkNextEntryAndCreateToken`, in that order), and the object doesn't really make sense until all that has been done.

We could enforce this temporal dependency by moving the other three methods inside `create`, which would then take the iterator, the previous token, and the batch size.